tiledb/sm/stats/stats.h

```diff
@@ -0,0 +1,355 @@
+    READ_COMPUTE_TILE_OVERLAP,
+    READ_COMPUTE_TILE_COORDS,
+    READ_INIT_STATE,
+    READ_NEXT_PARTITION,
+    READ_SPLIT_CURRENT_PARTITION,
+    READ,
+    DBG,
+    WRITE,
+    WRITE_SPLIT_COORDS_BUFF,
+    WRITE_CHECK_COORD_OOB,
+    WRITE_SORT_COORDS,
+    WRITE_CHECK_COORD_DUPS,
+    WRITE_COMPUTE_COORD_DUPS,
+    WRITE_PREPARE_TILES,
+    WRITE_COMPUTE_COORD_META,
+    WRITE_FILTER_TILES,
+    WRITE_TILES,
+    WRITE_STORE_FRAG_META,
+    WRITE_FINALIZE,
+    WRITE_CHECK_GLOBAL_ORDER,
+    WRITE_INIT_TILE_ITS,
+    WRITE_COMPUTE_CELL_RANGES,
+    WRITE_PREPARE_AND_FILTER_TILES,
+    WRITE_ARRAY_META,
+  };
+
+  /** Enumerates the stat counter types. */
+  enum class CounterType {
+    READ_ARRAY_SCHEMA_SIZE,
+    CONSOLIDATED_FRAG_META_SIZE,
+    READ_FRAG_META_SIZE,
+    READ_RTREE_SIZE,
+    READ_TILE_OFFSETS_SIZE,
+    READ_TILE_VAR_OFFSETS_SIZE,
+    READ_TILE_VAR_SIZES_SIZE,
+    READ_ARRAY_META_SIZE,
+    READ_NUM,
+    READ_BYTE_NUM,
+    READ_UNFILTERED_BYTE_NUM,
+    READ_ATTR_NUM,
+    READ_ATTR_FIXED_NUM,
+    READ_ATTR_VAR_NUM,
+    READ_DIM_NUM,
+    READ_DIM_FIXED_NUM,
+    READ_DIM_VAR_NUM,
+    READ_DIM_ZIPPED_NUM,
+    READ_OVERLAP_TILE_NUM,
+    READ_RESULT_NUM,
+    READ_CELL_NUM,
+    READ_LOOP_NUM,
+    WRITE_NUM,
+    WRITE_ATTR_NUM,
+    WRITE_ATTR_FIXED_NUM,
+    WRITE_ATTR_VAR_NUM,
+    WRITE_DIM_NUM,
+    WRITE_DIM_FIXED_NUM,
+    WRITE_DIM_VAR_NUM,
+    WRITE_DIM_ZIPPED_NUM,
+    WRITE_BYTE_NUM,
+    WRITE_FILTERED_BYTE_NUM,
+    WRITE_RTREE_SIZE,
+    WRITE_TILE_OFFSETS_SIZE,
+    WRITE_TILE_VAR_OFFSETS_SIZE,
+    WRITE_TILE_VAR_SIZES_SIZE,
+    WRITE_FRAG_META_FOOTER_SIZE,
+    WRITE_ARRAY_SCHEMA_SIZE,
+    WRITE_TILE_NUM,
+    WRITE_CELL_NUM,
+    WRITE_ARRAY_META_SIZE,
+  };
+
+  /* ****************************** */
+  /*   CONSTRUCTORS & DESTRUCTORS   */
+  /* ****************************** */
+
+  /** Constructor. */
+  Stats();
+
+  /** Destructor. */
+  ~Stats() = default;
+
+  /* ****************************** */
+  /*              API               */
+  /* ****************************** */
+
+  /** Adds `count` to the input timer stat. */
+  void add_timer(TimerType stat, double count);
+
+  /** Adds `count` to the input counter stat. */
+  void add_counter(CounterType stat, uint64_t count);
+
+  /** Returns true if statistics are currently enabled. */
+  bool enabled() const;
+
+  /** Enable or disable statistics gathering. */
+  void set_enabled(bool enabled);
+
+  /** Reset all counters to zero. */
+  void reset();
+
+  /** Dump the current stats to the given file. */
+  void dump(FILE* out) const;
+
+  /** Dump the current stats to the given string. */
+  void dump(std::string* out) const;
+
+ private:
+  /* ****************************** */
+  /*       PRIVATE ATTRIBUTES       */
+  /* ****************************** */
+
+  /** True if stats are being gathered. */
+  bool enabled_;
+
+  /** A map of timer stats, measured in seconds. */
+  std::map<TimerType, double> timer_stats_;
```

Comment:
I'll create a ticket to swap this `map` and the following to `unordered_map` instances, not urgent.